### PR TITLE
Add support for "mixed" IPv6 notation

### LINKED ISSUE #12: Support 'mixed' IPv6 representations
[RFC4291 specifies a "mixed" IPv6 syntax](https://tools.ietf.org/html/rfc4291#page-4) that is not currently supported:

> 3. An alternative form that is sometimes more convenient when dealing
>      with a mixed environment of IPv4 and IPv6 nodes is
>      `x:x:x:x:x:x:d.d.d.d`, where the '`x`'s are the hexadecimal values of
>      the six high-order 16-bit pieces of the address, and the '`d`'s are
>      the decimal values of the four low-order 8-bit pieces of the
>      address (standard IPv4 representation).  Examples:
>
>         `0:0:0:0:0:0:13.1.68.3`
>
>         `0:0:0:0:0:FFFF:129.144.52.38`
>
>      or in compressed form:
>
>         `::13.1.68.3`
>
>         `::FFFF:129.144.52.38`

### PR DESCRIPTION
**Pull request by Porges**
This fixes #12.

* Updated parser to accept mixed notation.
* Output mixed notation when encoding IPv4-mapped IPv6 addresses.

When parsing decimal-only IPv6 addresses, the parser is a little slower than the previous one (since it tries to parse an IPv4 address at each point) but it's very close. The parser is faster than previous one on hex-only addresses. Real life will be somewhere in between, but hopefully towards the latter 🙂 

Criterion output for old/new parser:
```
benchmarking IPv6 from Text/Old '::'
time                 36.15 ns   (35.96 ns .. 36.32 ns)
                     1.000 R   (1.000 R .. 1.000 R)
mean                 36.12 ns   (35.97 ns .. 36.35 ns)
std dev              607.9 ps   (375.0 ps .. 1.035 ns)
variance introduced by outliers: 22% (moderately inflated)

benchmarking IPv6 from Text/New '::'
time                 82.63 ns   (82.06 ns .. 83.24 ns)
                     1.000 R   (0.999 R .. 1.000 R)
mean                 82.52 ns   (82.13 ns .. 83.57 ns)
std dev              1.947 ns   (961.9 ps .. 3.687 ns)
variance introduced by outliers: 35% (moderately inflated)

benchmarking IPv6 from Text/Old '1:2:3:4:5:6:7:8'
time                 918.6 ns   (911.4 ns .. 929.6 ns)
                     0.999 R   (0.998 R .. 1.000 R)
mean                 913.6 ns   (907.7 ns .. 926.7 ns)
std dev              29.89 ns   (13.85 ns .. 49.71 ns)
variance introduced by outliers: 45% (moderately inflated)

benchmarking IPv6 from Text/New '1:2:3:4:5:6:7:8'
time                 987.1 ns   (981.0 ns .. 991.6 ns)
                     1.000 R   (1.000 R .. 1.000 R)
mean                 977.6 ns   (971.8 ns .. 984.2 ns)
std dev              20.63 ns   (16.52 ns .. 27.63 ns)
variance introduced by outliers: 25% (moderately inflated)

benchmarking IPv6 from Text/Old '1:2::7:8'
time                 546.5 ns   (542.8 ns .. 549.9 ns)
                     1.000 R   (0.999 R .. 1.000 R)
mean                 545.3 ns   (542.6 ns .. 549.1 ns)
std dev              10.58 ns   (6.913 ns .. 18.01 ns)
variance introduced by outliers: 24% (moderately inflated)

benchmarking IPv6 from Text/New '1:2::7:8'
time                 567.9 ns   (564.7 ns .. 571.6 ns)
                     1.000 R   (1.000 R .. 1.000 R)
mean                 570.2 ns   (568.3 ns .. 572.4 ns)
std dev              6.368 ns   (4.469 ns .. 9.211 ns)

benchmarking IPv6 from Text/Old 'a:b::c:d'
time                 548.5 ns   (545.9 ns .. 551.1 ns)
                     1.000 R   (0.999 R .. 1.000 R)
mean                 545.4 ns   (541.8 ns .. 551.2 ns)
std dev              14.06 ns   (9.531 ns .. 23.49 ns)
variance introduced by outliers: 35% (moderately inflated)

benchmarking IPv6 from Text/New 'a:b::c:d'
time                 445.4 ns   (443.0 ns .. 448.2 ns)
                     0.999 R   (0.999 R .. 1.000 R)
mean                 447.6 ns   (443.5 ns .. 458.2 ns)
std dev              20.45 ns   (9.651 ns .. 38.76 ns)
variance introduced by outliers: 64% (severely inflated)
```

### CODE REVIEW

File: test/Bench.hs [hunk 1]
```diff
@@ -19,6 +21,10 @@ main = do
   let ipAddr = IPv4 1000000009
       ipText = Text.pack "192.168.5.99"
       mac = Mac.fromOctets 0xFA 0xBB 0x43 0xA1 0x22 0x09
+      ip6Text = Text.pack ""
```
Comment (Porges) [inline test/Bench.hs]: doh

**Comment by andrewthad** [comment]
Thanks for doing this and for putting together tests for it. Fix the line in the benchmark suite that you commented on. I cannot push to your branch, but if you run `git pull https://github.com/andrewthad/haskell-ip mixed-ipv6`, you'll get some docs for `encode` and `decode` that I added in 8d66a5b4ed450ed619f26703b36b868afda97b52. Other than that, this looks good.

**Comment by andrewthad** [comment]
This looks good. Are you ready for this to be merged?

**Comment by Porges** [comment]
Do you need me to do anything about the version number?

**Comment by andrewthad** [comment]
Nah, I've got some other stuff I'm going to add before a release, so I'm going to do a major version bump soon anyway. Thanks!

**Comment by andrewthad** [comment]
This has been released in `ip-1.3.0`.

**Pull request merged**